extern/sector-storage/stores/remote.go

```diff
@@ -415,4 +419,178 @@ func (r *Remote) FsStat(ctx context.Context, id ID) (fsutil.FsStat, error) {
+	ft := storiface.FTUnsealed
+
+	// check if we have the unsealed sector file locally
+	paths, _, err := r.local.AcquireSector(ctx, s, ft, storiface.FTNone, storiface.PathStorage, storiface.AcquireMove)
+	if err != nil {
+		return nil, xerrors.Errorf("acquire local: %w", err)
+	}
+
+	path := storiface.PathByType(paths, ft)
+
+	if path != "" {
+		// if we have the unsealed file locally, return a reader that can be used to read the contents of the
+		// unsealed piece.
+		log.Infof("Read local %s (+%d,%d)", path, offset, size)
+		ssize, err := s.ProofType.SectorSize()
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("fetched sector size %s (+%d,%d)", path, offset, size)
+
+		// open the unsealed sector file for the given sector size located at the given path.
+		pf, err := r.pfHandler.OpenPartialFile(abi.PaddedPieceSize(ssize), path)
+		if err != nil {
+			return nil, xerrors.Errorf("opening partial file: %w", err)
+		}
+		log.Debugf("local partial file opened %s (+%d,%d)", path, offset, size)
+
+		// even though we have an unsealed file for the given sector, we still need to determine if we have the unsealed piece
+		// in the unsealed sector file. That is what `HasAllocated` checks for.
+		has, err := r.pfHandler.HasAllocated(pf, storiface.UnpaddedByteIndex(offset.Unpadded()), size.Unpadded())
+		if err != nil {
+			return nil, xerrors.Errorf("has allocated: %w", err)
+		}
+		log.Debugf("check if partial file is allocated %s (+%d,%d)", path, offset, size)
+
+		if !has {
+			log.Debugf("miner has unsealed file but not unseal piece, %s (+%d,%d)", path, offset, size)
+			if err := r.pfHandler.Close(pf); err != nil {
+				return nil, xerrors.Errorf("close partial file: %w", err)
+			}
+			return nil, nil
+		}
+
+		log.Infof("returning piece reader for local unsealed piece sector=%+v, (offset=%d, size=%d)", s.ID, offset, size)
+		return r.pfHandler.Reader(pf, storiface.PaddedByteIndex(offset), size)
+	}
+
+	// --- We don't have the unsealed sector file locally
+
+	// if we don't have the unsealed sector file locally, we'll first lookup the Miner Sector Store Index
+	// to determine which workers have the unsealed file and then query those workers to know
+	// if they have the unsealed piece in the unsealed sector file.
+	si, err := r.index.StorageFindSector(ctx, s.ID, ft, 0, false)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(si) == 0 {
+		return nil, xerrors.Errorf("failed to read sector %v from remote(%d): %w", s, ft, storiface.ErrSectorNotFound)
+	}
+
+	sort.Slice(si, func(i, j int) bool {
+		return si[i].Weight > si[j].Weight
+	})
+
+	for _, info := range si {
+		for _, url := range info.URLs {
+			// checkAllocated makes a JSON RPC query to a remote worker to determine if it has
+			// unsealed piece in their unsealed sector file.
+			ok, err := r.checkAllocated(ctx, url, s.ProofType, offset, size)
+			if err != nil {
+				log.Warnw("check if remote has piece", "url", url, "error", err)
```

Comment:
I'd suggest we keep track of the last error here.
If we're not able to get the unsealed piece from a remote worker, and there was an error querying a worker, we should return the last error.

Reply:
done.